File: test/api/cpp/finite_field.cpp
/******************************************************************************
 * Top contributors (to current version):
 *   Alex Ozdemir
 *
 * This file is part of the cvc5 project.
 *
 * Copyright (c) 2009-2022 by the authors listed in the file AUTHORS
 * in the top-level source directory and their institutional affiliations.
 * All rights reserved.  See the file COPYING in the top-level source
 * directory for licensing information.
 * ****************************************************************************
 *
 * An example of solving finite field problems with cvc5's cpp API
 */

#include <cassert>
#include <iostream>

#include "api/cpp/cvc5.h"

using namespace std;
using namespace cvc5;

int main()
{
  Solver solver;
  solver.setOption("produce-models", "true");

  // Make single precision floating-point variables
  Sort f5 = solver.mkFiniteFieldSort("5");
  Term a = solver.mkConst(f5, "a");
  Term b = solver.mkConst(f5, "b");
  Term z = solver.mkFiniteFieldElem("0", f5);

  Term inv =
      solver.mkTerm(EQUAL,
                    {solver.mkTerm(FINITE_FIELD_ADD,
                                   {solver.mkTerm(FINITE_FIELD_MULT, {a, b}),
                                    solver.mkFiniteFieldElem("-1", f5)}),
                     z});
  Term aIsTwo = solver.mkTerm(
      EQUAL,
      {solver.mkTerm(FINITE_FIELD_ADD, {a, solver.mkFiniteFieldElem("-2", f5)}),
       z});
  solver.assertFormula(inv);
  solver.assertFormula(aIsTwo);

  Result r = solver.checkSat();  // result is sat
  assert(r.isSat());

  cout << "a = " << solver.getValue(a) << endl;
  cout << "b = " << solver.getValue(b) << endl;

  Term bIsTwo = solver.mkTerm(
      EQUAL,
      {solver.mkTerm(FINITE_FIELD_ADD, {b, solver.mkFiniteFieldElem("-2", f5)}),
       z});

  solver.assertFormula(bIsTwo);

  r = solver.checkSat();  // result is sat
  assert(!r.isSat());
}

File: test/api/python/finite_field.py
###############################################################################
# Top contributors (to current version):
#   Alex Ozdemir
#
# This file is part of the cvc5 project.
#
# Copyright (c) 2009-2023 by the authors listed in the file AUTHORS
# in the top-level source directory and their institutional affiliations.
# All rights reserved.  See the file COPYING in the top-level source
# directory for licensing information.
# #############################################################################
#
# A simple start-up/tear-down test for cvc5.
#
# This simple test just makes sure that the public interface is
# minimally functional.  It is useful as a template to use for other
# system tests.
##

import cvc5
from cvc5 import Kind

slv = cvc5.Solver()
slv.setOption("produce-models", "true")
F = slv.mkFiniteFieldSort(5)
a = slv.mkConst(F, "a")
b = slv.mkConst(F, "b")
assert 5 == F.getFiniteFieldSize()

inv = slv.mkTerm(
    Kind.EQUAL,
    slv.mkTerm(Kind.FINITE_FIELD_MULT, a, b),
    slv.mkFiniteFieldElem(1, F),
)
aIsTwo = slv.mkTerm(
    Kind.EQUAL,
    a,
    slv.mkFiniteFieldElem(2, F),
)
slv.assertFormula(inv)
slv.assertFormula(aIsTwo)
r = slv.checkSat()
assert r.isSat()
assert slv.getValue(a).toPythonObj() == 2
assert slv.getValue(b).toPythonObj() == -2

bIsTwo = slv.mkTerm(
    Kind.EQUAL,
    b,
    slv.mkFiniteFieldElem(2, F),
)
slv.assertFormula(bIsTwo)
r = slv.checkSat()
assert not r.isSat()
